dask/dataframe/operation.py

```diff
@@ -0,0 +1,402 @@
+        return replace(self, _meta=self.meta.copy())
+
+    def __hash__(self):
+        return hash(tokenize(self.name, self.meta, self.divisions))
+
+
+@dataclass(frozen=True)
+class FrameCreation(FusableOperation, FrameOperation):
+
+    _io_func: Callable
+    _meta: Any
+    _inputs: list
+    _divisions: tuple
+    _label: str
+
+    @cached_property
+    def name(self) -> str:
+        token = tokenize(self._io_func, self._meta, self._inputs, self._divisions)
+        return f"{self._label}-{token}"
+
+    @cached_property
+    def dependencies(self) -> frozenset[CollectionOperation]:
+        _dep = LiteralInputs({(i,): val for i, val in enumerate(self._inputs)})
+        return frozenset({_dep})
+
+    @property
+    def meta(self) -> Any:
+        return self._meta
+
+    @cached_property
+    def divisions(self) -> tuple | None:
+        return tuple(self._divisions or (None,) * (len(self._inputs) + 1))
+
+    @cached_property
+    def subgraph_callable(
+        self,
+    ) -> tuple[SubgraphCallable, frozenset[CollectionOperation]]:
+        inkeys = [d.name for d in self.dependencies]
+        subgraph = {self.name: (self._io_func, inkeys[-1])}
+        return (
+            SubgraphCallable(
+                dsk=subgraph,
+                outkey=self.name,
+                inkeys=inkeys,
+            ),
+            self.dependencies,
+        )
+
+    def subgraph(self, keys: list[tuple]) -> tuple[dict, dict]:
+
+        # Always fuse MapInput operations
+        input_op = next(iter(self.dependencies))
+        input_op_name = input_op.name
+        input_op_keys = [(input_op_name,) + tuple(key[1:]) for key in keys]
+        dep_subgraph, _ = input_op.subgraph(input_op_keys)
+
+        # Build subgraph with LiteralInputs dependencies fused
+        dsk = {}
+        for key in keys:
+            dep_key = (input_op_name,) + tuple(key[1:])
+            dsk[key] = (self._io_func, dep_subgraph.get(dep_key, dep_key))
+        return dsk, {}
+
+    def reinitialize(
+        self, replace_dependencies: dict[str, CollectionOperation], **changes
+    ) -> FrameCreation:
+        # TODO: Support column projection and predicate pushdown
+        return replace(self, **changes)
+
+    def __hash__(self):
+        return hash(self.name)
+
+
+@dataclass(frozen=True)
+class PartitionwiseOperation(FusableOperation, FrameOperation):
+
+    _func: Callable
+    _meta: Any
+    _args: list[Any]
+    _divisions: tuple
+    _label: str
+    _kwargs: dict
+
+    @cached_property
+    def name(self) -> str:
+        token = tokenize(
+            self._func, self._meta, self._args, self._divisions, self._kwargs
+        )
+        return f"{self._label}-{token}"
+
+    @property
+    def meta(self) -> Any:
+        return self._meta
+
+    def replace_meta(self, value) -> PartitionwiseOperation:
+        return replace(self, _meta=value)
+
+    @property
+    def divisions(self) -> tuple | None:
+        return self._divisions
```

Comment:
Do we need to allow `divisions` to be `None`? That doesn't fly with [`check_divisions`](https://github.com/dask/dask/blob/23f6174a5e4fad8edc9cb22a966e7800ff41f571/dask/dataframe/core.py#L6978-L6988), and it would simplify the interface for this and `npartitions`

Reply:
There **are** some places where divisions can be None, but I think adding a ScalarOperation would probably remove most of these cases. Overall, I agree that removing the None option is probably the way to go.